src/relationUtils.ts

```diff
@@ -276,3 +280,118 @@ export function toJson(output: Relation[] | ArrowRelation[]): any {
 
   return result;
 }
+
+export function filterRelations(relations: ArrowRelation[], keys: string[]) {
+  const len = keys.length;
+  const keysStr = keys.join('/');
+  const filteredRelations: ArrowRelation[] = [];
+
+  relations.forEach(r => {
+    const relationKeys = getKeys(r.relationId);
+
+    if (relationKeys.slice(0, len).join('/') === keysStr) {
+      const newKeys = relationKeys.slice(len);
+
+      filteredRelations.push({
+        ...r,
+        relationId: getRelationId(newKeys),
+      });
+    }
+  });
+
+  return filteredRelations;
+}
+
+export type Pos = {
+  line: number;
+  character: number;
+};
+
+export type Range = {
+  start: Pos;
+  end: Pos;
+};
+
+export type Diagnostic = {
+  code: string;
+  message: string;
+  severity: 'exception' | 'error' | 'warning' | 'info' | 'suggestion';
+  report?: string;
+  range?: Range[];
+  model?: string;
+};
+
+export function parseDiagnostics(relations: ArrowRelation[]) {
+  relations = filterRelations(relations, [':rel', ':catalog', ':diagnostic']);
+
+  const diagnostics: Diagnostic[] = [];
+
+  setRangeProp(
+    ['start', 'line'],
+    diagnostics,
+    filterRelations(relations, [':range', ':start', ':line'])[0],
```

Comment:
why the `[0]` indexing here? what if more than one relation matches the filter keys?

Reply:
There can't be more than one `/:rel/:catalog/:diagnostic/:range/:start/:line/*`. Right?

Like the whole parsing function is based on the assumption that there're 4 relations for the range.

Reply:
There can be multiple, both iterating on the diagnostics (first index), but also the range (second index)

```scala
def foo = 1,2,3
    def bar = 4,5,6
    // x... has arities 2 and 1
    def baz = x... : foo(x..., 3) and bar(4, 5, x...)
```

has a diagnostic with two ranges for instance

Reply:
I might be confusing relation and tuple here.  There is only one relation prefixed by `/:rel/:catalog/:diagnostic/:range/:start/:line`, but it can contain multiple tuples.

Reply:
Oh yeah, I figured that. It's handled. See the test https://github.com/RelationalAI/rai-sdk-javascript/pull/18/files#diff-359af596685a75d0adbc6f90a969e0b3fbb842c2dc70cc129bfc03851b92c083R105

There're two ranges in the first diagnostic.